platform/macos/joypad_macos.mm

```diff
@@ -30,303 +30,275 @@
 
 #import "joypad_macos.h"
 
-#include <Foundation/Foundation.h>
-
-#import "os_macos.h"
+#include <CoreHaptics/CoreHaptics.h>
+#import <os/log.h>
 
 #include "core/config/project_settings.h"
-#include "core/os/keyboard.h"
 #include "core/string/ustring.h"
 #include "main/main.h"
 
-@implementation RumbleMotor
-
-- (instancetype)initWithController:(GCController *)controller locality:(GCHapticsLocality)locality {
-	self = [super init];
-	self.engine = [controller.haptics createEngineWithLocality:locality];
-	self.player = nil;
-	return self;
-}
-
-- (void)execute_pattern:(CHHapticPattern *)pattern {
-	NSError *error;
-	id<CHHapticPatternPlayer> player = [self.engine createPlayerWithPattern:pattern error:&error];
-
-	// When all players have stopped for an engine, stop the engine.
-	[self.engine notifyWhenPlayersFinished:^CHHapticEngineFinishedAction(NSError *_Nullable error) {
-		return CHHapticEngineFinishedActionStopEngine;
-	}];
-
-	self.player = player;
+class API_AVAILABLE(macos(11), ios(13.0), tvos(14.0)) RumbleMotor {
+	CHHapticEngine *engine;
+	id<CHHapticPatternPlayer> player;
+	bool is_started;
 
-	// Starts the engine and returns if an error was encountered.
-	if (![self.engine startAndReturnError:&error]) {
-		print_verbose("Couldn't start controller haptic engine");
-		return;
-	}
-	if (![self.player startAtTime:0 error:&error]) {
-		print_verbose("Couldn't execute controller haptic pattern");
+	RumbleMotor(GCController *p_controller, GCHapticsLocality p_locality) {
+		engine = [p_controller.haptics createEngineWithLocality:p_locality];
+		engine.autoShutdownEnabled = YES;
```

Comment:
Haptics was buggy, as the previous implementation was continually stopping and starting the haptic engine. When playing haptics rapidly, requiring the existing player to stop, would ultimately cause the engine to enter an inconsistent state an haptics would stop working. This PR utilises the [`autoShutdownEnabled`](https://developer.apple.com/documentation/corehaptics/chhapticengine/isautoshutdownenabled?language=objc) feature to let the OS handle it automatically.